luigi/task.py

```diff
@@ -47,27 +47,72 @@
 TASK_ID_TRUNCATE_PARAMS = 16
 TASK_ID_TRUNCATE_HASH = 10
 TASK_ID_INVALID_CHAR_REGEX = re.compile(r'[^A-Za-z0-9_]')
+_SAME_AS_PYTHON_MODULE = '_same_as_python_module'
 
 
-def namespace(namespace=None):
+def namespace(namespace=None, scope=''):
     """
     Call to set namespace of tasks declared after the call.
 
-    It is best practice to call this function without arguments at the end of any file it has been
-    used in. That is to ensure that subsequent tasks have the default namespace again.
+    It is often desired to call this function with the keyword argument
+    ``scope=__name__``.
+
+    The ``scope`` keyword makes so this call is only affective for task classes
+    with a matching [*]_ ``__module__``. The default value for ``scope`` is the
+    empty string, which means all classes. Multiple calls with the same scope
+    simply replace each other.
 
     The namespace of a :py:class:`Task` can also be changed by specifying the property
-    ``task_namespace``. This solution has the advantage that the namespace
-    doesn't have to be restored.
+    ``task_namespace``.
 
     .. code-block:: python
 
         class Task2(luigi.Task):
             task_namespace = 'namespace2'
 
+    This explicit setting takes priority over whatever is set in
+    ``namespace()`` method, and it's also inherited through normal python
+    inheritence.
+
     There's no equivalent way to set the ``task_family``.
+
+    *New since Luigi 2.6.0:* ``scope`` keyword argument.
+
+    .. [*] When there are multiple levels of matching module scopes like
+           ``a.b`` vs ``a.b.c``, the more specific one wins.
+    .. seealso:: The new and better scaling :py:func:`auto_namespace`
     """
-    Register._default_namespace = namespace or ''
+    Register._default_namespace_dict[scope] = namespace or ''
+
+
+def auto_namespace(scope=''):
+    """
+    Same as :py:func:`namespace`, but instead of a constant namespace, it will
+    be set to the ``__module__`` of the task class. This is desirable for these
+    reasons:
+
+     * Two tasks with the same name will not have conflicting task families
+     * It's more pythonic, as modules are pythons recommended way to
+       do namespacing.
+     * It's traceable, when you see the full name of a task, you can immedietly
```

Comment:
It's traceable. __*When*__ you see the full name of a task, you can __*immediately*__